[Torch-to-Linalg] Add dynamic dimension support for BroadcastTo op

Abhishek-Varma opened the pull request:
-- This commit adds support for dynamic dimension in BroadcastTo op.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

Abhishek-Varma commented:
Hi @ramiro050 could you please review this PR when you get a chance?

ramiro050 commented on lib/Conversion/TorchToLinalg/Utils.cpp:
```diff
@@ -335,14 +336,18 @@ LogicalResult torch_to_linalg::broadcastToGivenShape(
+  Value oneIndex =
+      rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
   size_t diff = broadcastToShape.size() - inputShape.size();
+  bool shiftToBroadcastShapeEmpty = shiftToBroadcastShape.size() == 0;
```
nit: unnecessary variable. You can use `shiftToBroadcastShape.empty()`

ramiro050 commented on python/torch_mlir_e2e_test/test_suite/basic.py:
```diff
@@ -1333,6 +1333,33 @@ def BroadcastZeroRankInputStaticModule_basic(module, tu: TestUtils):
+        dim_at_index_1 = torch.ops.aten.size(x, 1)
+        dim_at_index_3 = torch.ops.aten.size(x, 3)
+        res = torch.broadcast_to(y, [1, dim_at_index_1, 1, dim_at_index_3])
+        # print(res)
```
remove print statement

ramiro050 commented on python/torch_mlir_e2e_test/test_suite/basic.py:
```diff
@@ -1333,6 +1333,33 @@ def BroadcastZeroRankInputStaticModule_basic(module, tu: TestUtils):
+    def forward(self, x, y):
+        dim_at_index_1 = torch.ops.aten.size(x, 1)
+        dim_at_index_3 = torch.ops.aten.size(x, 3)
+        res = torch.broadcast_to(y, [1, dim_at_index_1, 1, dim_at_index_3])
```
use `torch.ops.aten.broadcast_to`

ramiro050 commented on lib/Conversion/TorchToLinalg/DataMovement.cpp:
```diff
@@ -1179,12 +1179,26 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
       return rewriter.notifyMatchFailure(
           op, "unimplemented: the size list is not from list construct");
     }
+    SmallVector<bool> shiftToBroadcastShape;
```
I'm not sure I understand what this represents. Can you explain to me the reasoning?

Abhishek-Varma replied:
Hi @ramiro050 ,

1. I've now renamed this to `useBroadcastToShape`.
2. `broadcastToShape` is already a construct which is/was previously being used in the lowering.
3. Now with `useBroadcastToShape` I'm trying to use `broadcastToShape` only for those dimensions which are dynamic and whose final shape needs to be inferred based on the "shape being broadcastedTo".

I've added a comment too in the file to address this.

Let me know if that sounds good to you.

Abhishek-Varma commented:
Hi @ramiro050 

I've addressed your comments so far.
Can you take a look at the PR and re-review?

Also, the git checkout of submodules look like an orthogonal bug? @powderluv

powderluv commented:
@ramiro050 is on vacation so please get someone else to review 

I will fix the CI

vivekkhandelwal1 reviewed (approved):
LGTM

Abhishek-Varma commented:
Hi @powderluv 

Can we get this in and address any review comments (if any) later in a subsequent patch?
